rust/otap-dataflow/docs/tracing-proposal.md

```diff
@@ -0,0 +1,145 @@
+diagnosing data loss, performance bottlenecks, misrouted signals, stateful
+processing delays, or the impact of control-plane actions becomes challenging.
+
+**Traditional metrics and logs alone are insufficient** for root-cause analysis,
+end-to-end flow visibility, and operational insights, especially under dynamic
+configuration and sampling regimes.
+
+**Goal**:
+Leverage OpenTelemetry tracing primitives (traces, spans, events, links)
+to enable robust debugging, troubleshooting, monitoring, and optimization of the
+entire dataflow engine, **without imposing excessive overhead or losing critical
+information through sampling.**
+
+## System Overview
+
+The dataflow engine is structured as a DAG of the following node types:
+
+- **Receivers**: Ingest and translate data from external sources (may have
+  multiple output ports).
+- **Processors**: Transform, filter, batch, or enrich data (may have multiple
+  output ports).
+- **Exporters**: Deliver processed data to external systems.
+
+Each node features:
+
+- A **unique id** and **type** (i.e. urn).
+- A **control channel** for orchestration commands (e.g. config_update, ack,
+  timer_tick, shutdown).
+- A **pdata channel** (processors/exporters) for ingesting batches of signals (
+  metrics, logs, events, spans).
+
+**PData messages** are batches containing a single signal type.
+
+## Tracing Modes
+
+- **No Tracing**: No spans emitted.
+- **Tail Sampling**: Only selected traces (e.g. errors, random sample) are
+  exported, but all nodes are instrumented.
+- **Full Tracing**: Every dataflow activity is traced.
+
+> **Metrics are derived from spans before sampling**, ensuring high-fidelity
+> monitoring regardless of trace sampling.
+
+## Mapping OTEL Tracing Primitives to the DAG
+
+**Trace**
+
+- Definition: Represents a single PData flow journey through the DAG.
+- Purpose: Enables full end-to-end flow and performance visibility.
+
+**Span**
+
+- Definition: Captures a node’s processing of a PData flow.
+- Attributes:
+    - node.id, node.type, signal.type
+    - pdata.batch.size, out.port
+    - stateful, latency.ms, error
+- Purpose: Provides detailed, per-node insight into processing time, routing,
+  state, and errors.
+
+**Span Events**
+
+- Definition: Record significant in-node occurrences.
+- Examples:
+    - Control message handling (ack, time_tick, shutdown)
+    - State transitions (batch full/flush/drop)
+    - Output port selection, backpressure, errors
+
+**Span Links**
+
+- Definition: Capture relationships between spans when batches are split (
+  fan-out) or merged (fan-in) across nodes.
+- Purpose: Enables lineage and parallel flow reconstruction.
+
+**Control Plane Integration**
+
+- Control actions generate events or standalone spans, linked to affected
+  data spans as needed, providing operational audit trails.
+
+**Channel Utilization Tracking**
+
+- Utilization Metrics (derived from span events/attributes or direct metrics):
```

Comment:
what is the general reasoning behind leverage span to derive metrics instead of always using direct metrics? In the event Spans prove expensive, and a decision to do head-based sampling is made, the metrics would get inaccurate...

Reply:
I will clarify this point in the doc.

Reply:
We should focus on the instrumentation points -- the "what happens when" question should be configurable behavior determined by the SDK which can derive metrics or create spans. (Again: is this the new OTAP SDK? 😀)

Also: we can use sampling to derive metrics from sampled events using modern OTel specifications, and I think this project would make a good showcase Span-to-Metrics pipeline. We should be able to sample SDK events into an OTAP frame, apply logic to those spans (filter, project, etc), then derive metrics from sampling-adjusted counts into another OTAP frame.